client/factory.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	gocache "github.com/patrickmn/go-cache"
+
+	"github.com/giantswarm/azure-operator/v4/pkg/credential"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+const (
+	cacheHitLogKey       = "cacheHit"
+	clientTypeLogKey     = "clientType"
+	credentialNameLogKey = "credentialName"
+)
+
+type FactoryConfig struct {
+	CacheDuration time.Duration
+	GSTenantID    string
+	K8sClient     k8sclient.Interface
+	Logger        micrologger.Logger
+}
+
+// Factory is creating Azure clients for specified AzureConfig CRs, so basically for specified
+// tenant clusters. All created clients are cached.
+type Factory struct {
+	gsTenantID string
+	k8sClient  k8sclient.Interface
+	logger     micrologger.Logger
+	mutex      sync.Mutex
+
+	// map [credentialName + client type] -> client
+	cachedClients *gocache.Cache
+}
+
+type clientCreatorFunc func(autorest.Authorizer, string, string) (interface{}, error)
+
+// NewFactory returns a new Azure client factory that is used throughout entire azure-operator
+// lifetime.
+func NewFactory(config FactoryConfig) (*Factory, error) {
+	if config.CacheDuration < 5*time.Minute { // cache at least for one reconciliation loop duration
+		return nil, microerror.Maskf(invalidConfigError, "%T.CacheDuration must be at least 5 minutes", config)
+	}
+	if len(config.GSTenantID) == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "%T.GSTenantID must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	factory := &Factory{
+		k8sClient:     config.K8sClient,
+		logger:        config.Logger,
+		gsTenantID:    config.GSTenantID,
+		cachedClients: gocache.New(config.CacheDuration, 2*config.CacheDuration),
+	}
+
+	factory.cachedClients.OnEvicted(func(clientKey string, i interface{}) {
+		factory.onEvicted(clientKey)
+	})
+
+	return factory, nil
+}
+
+// GetDeploymentsClient returns DeploymentsClient that is used for management of deployments and
+// ARM templates. The client (for specified cluster) is cached after creation, so the same client
+// is returned every time.
+func (f *Factory) GetDeploymentsClient(cr v1alpha1.AzureConfig) (*resources.DeploymentsClient, error) {
+	createClientFunc := func(authorizer autorest.Authorizer, subscriptionID string, partnerID string) (interface{}, error) {
+		return newDeploymentsClient(authorizer, subscriptionID, partnerID)
+	}
```

Comment:
Is this wrapping needed only because the return type differs between those two functions? i.e. `createClientFunc` returns `(interface{}, error)` whereas `newDeploymentsClient()` returns `(*resources.DeploymentsClient, error)`?

If so, would it make sense if you change the signature of `new*Client()` functions to return `interface{}` and introduce casting functions for each type? The casting function could be then also used together with `getClient()` since that also returns `interface{}`.

After that change this function could roughly look like this:
```go
client, err := f.getClient(cr, "DeploymentsClient", newDeploymentsClient)
if err != nil {
    return nil, microerror.Mask(err)
}

return toDeploymentsClient(client), err
```

If there is still old code using `newDeploymentsClient` as it used to be, you could wrap those with `toDeploymentsClient()`. It would maybe look a bit funny, but in this case the benefit would be on the new code and old maintained code is the one that suffers slight readability hit. WDYT?

Reply:
Ok doing this now, as I have to fix error handling anyway :)